base/src/commonMain/kotlin/flow/flow.kt

```diff
@@ -0,0 +1,118 @@
+/*
+ Copyright 2021 Splendo Consulting B.V. The Netherlands
+
+    Licensed under the Apache License, Version 2.0 (the "License");
+    you may not use this file except in compliance with the License.
+    You may obtain a copy of the License at
+
+      http://www.apache.org/licenses/LICENSE-2.0
+
+    Unless required by applicable law or agreed to in writing, software
+    distributed under the License is distributed on an "AS IS" BASIS,
+    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+    See the License for the specific language governing permissions and
+    limitations under the License.
+
+ */
+
+package com.splendo.kaluga.base.flow
+
+import com.splendo.kaluga.base.flow.SharedFlowCollectionEvent.NoMoreCollections
+import com.splendo.kaluga.base.flow.SharedFlowCollectionEvent.FirstCollection
+import com.splendo.kaluga.base.flow.SharedFlowCollectionEvent.LaterCollections
+import kotlinx.coroutines.flow.Flow
+import kotlinx.coroutines.flow.FlowCollector
+import kotlinx.coroutines.flow.MutableSharedFlow
+import kotlinx.coroutines.flow.collect
+import kotlinx.coroutines.flow.distinctUntilChanged
+import kotlinx.coroutines.flow.dropWhile
+import kotlinx.coroutines.flow.filterNot
+import kotlinx.coroutines.flow.map
+import kotlinx.coroutines.flow.transformWhile
+import kotlinx.coroutines.flow.withIndex
+
+enum class SharedFlowCollectionEvent {
+    NoMoreCollections, FirstCollection, LaterCollections
+}
+
+/**
+ * Calls back with events in the following order:
+ *
+ * [FirstCollection] -> [NoMoreCollections] -> [LaterCollections] -> back to [NoMoreCollections] when all flows are terminated
+ *
+ * If a flow already exists on this [MutableSharedFlow] the [FirstCollection] event is immediately raised.
+ * If no subscriptions exist
+ *
+ * Note that is this [MutableSharedFlow] was subscribed and unsubscribed before this method was called,
+ * a new subscription will still raise the [FirstCollection] event
+ */
+suspend fun <T> MutableSharedFlow<T>.onCollectionEvent(onEvent: suspend MutableSharedFlow<T>.(SharedFlowCollectionEvent) -> Unit): Unit =
+    subscriptionCount
+        .map { it > 0 } // simplify to zero or more-than-one
```

Comment:
Comment says to zero or more, but it's more than zero

Reply:
good catch, but the "or" refresh to 0 (`false`) _or_ more than zero (`true`), I'll clarify the comment.